dune/grid/CpGrid.hpp

```diff
@@ -863,6 +864,46 @@ namespace Dune
             return ret;
         }
 
+        template<class Cell2FacesRowIterator>
+        int
+        faceTag(const Cell2FacesRowIterator& cell_face) const
+        {
+            const int cell = cell_face.getCellIndex();
+            const int face = *cell_face;
+            assert (0 <= cell);  assert (cell < numCells());
+            assert (0 <= face);  assert (face < numFaces());
+
+            typedef cpgrid::OrientedEntityTable<1,0>::row_type F2C;
+
+            const cpgrid::EntityRep<1> f(face, true);
+            const F2C&     f2c = current_view_data_->face_to_cell_[f];
+            const face_tag tag = current_view_data_->face_tag_[f];
+
+            assert ((f2c.size() == 1) || (f2c.size() == 2));
+
+            const bool normal_is_in =
+                ((f2c.size() == 1)
+                 ? ! f2c[0].orientation() // Single cell => boundary
+                 : (f2c[0].index() != cell));     // Two cells => interior
+
+            switch (tag) {
+            case LEFT:
+                //                    LEFT : RIGHT
+                return normal_is_in ? 0    : 1; // min(I) : max(I)
+
+            case BACK:
+                //                    BACK : FRONT
+                return normal_is_in ? 1    : 2; // min(J) : max(J)
+
+            case TOP:
+                // Note: TOP at min(K) as 'z' measures *depth*.
+                //                    TOP  : BOTTOM
+                return normal_is_in ? 4    : 5; // min(K) : max(K)
+            default:
+                OPM_THROW(std::runtime_error, "Unhandeled face tag. This should never happen!");
```

Comment:
Minor typo: Unhandled. Also, I would call this a logic_error, not a runtime_error, since it will always be due to programmer and not user/input error.